x/autopilot/keeper/liquidstake.go

```diff
@@ -16,103 +15,141 @@ import (
 }
 
-func (k Keeper) RunLiquidStake(ctx sdk.Context, addr sdk.AccAddress, token sdk.Coin, packetMetadata types.StakeibcPacketMetadata) error {
+// Submits a LiquidStake message from the transfer receiver
+// If a forwarding recipient is specified, the stTokens are ibc transferred
+func (k Keeper) RunLiquidStake(
+	ctx sdk.Context,
+	amount sdkmath.Int,
+	transferMetadata transfertypes.FungibleTokenPacketData,
+	autopilotMetadata types.StakeibcPacketMetadata,
+) error {
 	msg := &stakeibctypes.MsgLiquidStake{
-		Creator:   addr.String(),
-		Amount:    token.Amount,
-		HostDenom: token.Denom,
+		Creator:   transferMetadata.Receiver,
+		Amount:    amount,
+		HostDenom: transferMetadata.Denom,
 	}
 
 	if err := msg.ValidateBasic(); err != nil {
 		return err
 	}
 
 	msgServer := stakeibckeeper.NewMsgServerImpl(k.stakeibcKeeper)
-	result, err := msgServer.LiquidStake(
+	msgResponse, err := msgServer.LiquidStake(
 		sdk.WrapSDKContext(ctx),
 		msg,
 	)
 	if err != nil {
 		return errorsmod.Wrapf(err, err.Error())
 	}
 
-	if packetMetadata.IbcReceiver == "" {
+	// If the IBCReceiver is empty, there is no forwarding step
+	if autopilotMetadata.IbcReceiver == "" {
 		return nil
 	}
 
-	hostZone, err := k.stakeibcKeeper.GetHostZoneFromHostDenom(ctx, token.Denom)
+	// Otherwise, if there is forwarding info, submit the IBC transfer
+	return k.IBCTransferStToken(ctx, msgResponse.StToken, transferMetadata, autopilotMetadata)
+}
+
+// Submits an IBC transfer of the stToken to a non-stride zone (either back to the host zone or to a different zone)
+// The sender of the transfer is the hashed receiver of the original autopilot inbound transfer
+func (k Keeper) IBCTransferStToken(
+	ctx sdk.Context,
+	stToken sdk.Coin,
+	transferMetadata transfertypes.FungibleTokenPacketData,
+	autopilotMetadata types.StakeibcPacketMetadata,
+) error {
+	hostZone, err := k.stakeibcKeeper.GetHostZoneFromHostDenom(ctx, transferMetadata.Denom)
 	if err != nil {
 		return err
 	}
 
-	return k.IBCTransferStAsset(ctx, result.StToken, addr.String(), hostZone, packetMetadata)
-}
-
-func (k Keeper) IBCTransferStAsset(ctx sdk.Context, stAsset sdk.Coin, sender string, hostZone *stakeibctypes.HostZone, packetMetadata types.StakeibcPacketMetadata) error {
-	ibcTransferTimeoutNanos := k.stakeibcKeeper.GetParam(ctx, stakeibctypes.KeyIBCTransferTimeoutNanos)
-	timeoutTimestamp := uint64(ctx.BlockTime().UnixNano()) + ibcTransferTimeoutNanos
-	channelId := packetMetadata.TransferChannel
+	// If there's no channelID specified in the packet, default to the channel on the host zone
+	channelId := autopilotMetadata.TransferChannel
 	if channelId == "" {
 		channelId = hostZone.TransferChannelId
 	}
+
+	// Generate a hashed address for the sender to the next hop,
+	// to prevent impersonation at downstream zones
+	// Note: The channel ID here is different than the one used in PFM
+	// (we use the outbound channelID, they use the inbound channelID)
+	// DOUBLE CHECK ME that it shouldn't matter
```

Comment:
Why do they include the channel in the hash in the first place? Not clear you need one at all. In fact, for our use case, there should only be one acceptable channel from host to Stride, so (check me) the channel doesn't add any information to the hash (it will always be the same for a given address).

Including the outbound channelID _does_ differentiate between LS-from-addrX&forward-to-chainA and LS-from-addrX&forward-to-chainB. Doesn't seem like we care whether there's a collision there but still might be nice to avoid it.

Reply:
Leaving this open until @asalzmann chimes in. Please feel free to close thereafter.

Reply:
By default, I'd opt to use the inbound channel id (given that's what PFM does). 

Actually, it's not entirely clear to me why they use the channel id in the has, so I'd be hesitant to make a decision on this before understanding that.

My understanding is as follows
* The bug in PFM was that anyone could forward via arbitrary addresses, leading to untrusted senders
* To resolve this, the high-level idea was to obfuscate senders by hashing them
* To implement this solution, the hash was hash(channel, sender)

Shouldn't hashing the sender alone obfuscate the sender for downstream chains?

@agouin would you be able to weigh in on why PFM _also_ includes the channel in the hash?

Reply:
Riley: Hey team we're shipping an upgrade to autopilot that hashes channel and sender (like PFM), but are considering removing the channel from the hash inputs. Tagged you guys in our PR here with more context: https://github.com/Stride-Labs/stride/pull/1038#discussion_r1443767344
Do you know why PFM includes the channel in the hash?

Aidan: My impression of the issue was that a previous version of PFM passed along untrusted senders, and the solution was to obfuscate the sender by hashing it. Doesn't hashing just the sender obfuscate it? Why do you also need the transfer channel?

Andrew: Just the sender should be enough afaict. We followed the same pattern as IBC hooks. I didn't see a problem with including the channel so I included it also.

It adds more uniqueness but the bech32 prefix of the source sender is included in the hash anyways

So it's mainly just extra collision prevention


**Based on this, would probably use the incoming channel id to stay consistent with PFM. 

But if it's a lot of extra work, it's not so important.**

Reply:
Based on that, I think I'd vote to leave it how we have it in this PR. Avoids having to pass an extra parameter across multiple functions and gives the same (weak) collision resistance as the inbound channel ID

Reply:
sgtm

Reply:
sgtm too!

Reply:
actually nvm, now that we're doing option 3, it's actually the same amount of work to use the same channel as PFM (and it's already implemented that way in the option 3 PR)